main.go

```diff
@@ -0,0 +1,175 @@
+
+		if response, err := http.Get(url); err != nil {
+			fmt.Printf("failed to fetch: %v\n", err)
+		} else if response.StatusCode != http.StatusOK {
+			fmt.Printf("failed to fetch: %s\n", http.StatusText(response.StatusCode))
+		} else {
+			defer response.Body.Close()
+			return ioutil.ReadAll(response.Body)
+		}
+
+		time.Sleep(delay)
+		delay *= 2
+		if delay > c.MaxBackoff {
+			delay = c.MaxBackoff
+		}
+	}
+
+	return nil, fmt.Errorf("timed out while fetching %q", url)
+}
+
+func main() {
+	flags := struct {
+		provider string
+		output   string
+		version  bool
+	}{}
+
+	flag.StringVar(&flags.provider, "provider", "", "The name of the cloud provider")
+	flag.StringVar(&flags.output, "output", "", "The file into which the metadata is written")
+	flag.BoolVar(&flags.version, "version", false, "Print the version and exit")
+
+	flag.Parse()
+
+	if flags.version {
+		fmt.Println(versionString)
+		return
+	}
+
+	switch flags.provider {
+	case "ec2":
+	default:
+		fmt.Fprintf(os.Stderr, "invalid provider %q\n", flags.provider)
+		os.Exit(2)
+	}
+
+	if err := os.MkdirAll(path.Dir(flags.output), 0755); err != nil {
+		fmt.Fprintf(os.Stderr, "failed to create directory: %v\n", err)
+		os.Exit(1)
+	}
+
+	out, err := os.Create(flags.output)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to create file: %v\n", err)
+		os.Exit(1)
+	}
+	defer out.Close()
+
+	if metadata, err := fetch_metadata(flags.provider); err == nil {
+		if err := write_metadata(out, metadata); err != nil {
+			fmt.Fprintf(os.Stderr, "failed to write metadata: %v\n", err)
+			os.Exit(1)
+		}
+	} else {
+		fmt.Fprintf(os.Stderr, "failed to fetch metadata: %v\n", err)
+		os.Exit(1)
+	}
+}
+
+func fetch_string(url string) (string, error) {
+	body, err := retryClient{
+		InitialBackoff: time.Second,
+		MaxBackoff:     time.Second * 5,
+		MaxAttempts:    10,
+	}.Get(url)
+	return string(body), err
+}
+
+func fetch_ip(url string) (net.IP, error) {
+	str, err := fetch_string(url)
+	if err != nil {
+		return nil, err
+	}
+	if ip := net.ParseIP(str); ip != nil {
+		return ip, nil
+	} else {
+		return nil, fmt.Errorf("couldn't parse %q as IP address", str)
+	}
+}
+
+func fetch_metadata(provider string) (metadata, error) {
+	switch provider {
+	case "ec2":
+		public, err := fetch_ip("http://169.254.169.254/2009-04-04/meta-data/public-ipv4")
+		if err != nil {
+			return metadata{}, err
+		}
+		local, err := fetch_ip("http://169.254.169.254/2009-04-04/meta-data/local-ipv4")
+		if err != nil {
+			return metadata{}, err
+		}
+		hostname, err := fetch_string("http://169.254.169.254/2009-04-04/meta-data/hostname")
+		if err != nil {
+			return metadata{}, err
+		}
+
+		return metadata{
+			PublicIPv4: public,
+			LocalIPv4:  local,
+			Hostname:   hostname,
+		}, nil
+	default:
+		panic("bad provider")
+	}
+}
+
+func write_metadata(out *os.File, metadata metadata) error {
+	if _, err := out.Write([]byte(fmt.Sprintf("COREOS_IPV4_PUBLIC=%s\n", metadata.PublicIPv4))); err != nil {
```

Comment:
Write is silly because: http://golang.org/pkg/os/#File.WriteString
But then Sprintf is silly because: http://golang.org/pkg/fmt/#Fprintf